File: sdk/python/aragora/namespaces/verticals.py
"""
Verticals Namespace API

Provides domain-specific vertical management:
- List and configure industry verticals
- Get vertical-specific tools and compliance
- Create specialized agents and debates

Features:
- Industry-specific configurations
- Compliance framework management
- Specialized agent creation
"""

from __future__ import annotations

from typing import TYPE_CHECKING, Any, Literal

if TYPE_CHECKING:
    from ..client import AragoraAsyncClient, AragoraClient


VerticalId = Literal[
    "software",
    "legal",
    "healthcare",
    "accounting",
    "finance",
    "research",
    "education",
    "marketing",
    "operations",
    "general",
]

ComplianceLevel = Literal["advisory", "warning", "enforced"]
ConsensusType = Literal["majority", "unanimous", "weighted"]


class VerticalsAPI:
    """
    Synchronous Verticals API.

    Provides methods for managing industry verticals:
    - List and configure verticals
    - Get vertical-specific tools and compliance frameworks
    - Create specialized agents and debates

    Example:
        >>> client = AragoraClient(base_url="https://api.aragora.ai", api_key="...")
        >>> verticals = client.verticals.list()
        >>> tools = client.verticals.get_tools("legal")
    """

    def __init__(self, client: AragoraClient):
        self._client = client

    def list(self, keyword: str | None = None) -> dict[str, Any]:
        """
        List all available verticals.

        Args:
            keyword: Optional keyword filter

        Returns:
            Dict with verticals list
        """
        params = {"keyword": keyword} if keyword else None
        return self._client.request("GET", "/api/verticals", params=params)

    def get(self, vertical_id: VerticalId) -> dict[str, Any]:
        """
        Get vertical details.

        Args:
            vertical_id: The vertical ID

        Returns:
            Dict with full vertical details
        """
        return self._client.request("GET", f"/api/verticals/{vertical_id}")

    def get_tools(self, vertical_id: VerticalId) -> dict[str, Any]:
        """
        Get tools available for a vertical.

        Args:
            vertical_id: The vertical ID

        Returns:
            Dict with tools list
        """
        return self._client.request("GET", f"/api/verticals/{vertical_id}/tools")

    def get_compliance(
        self,
        vertical_id: VerticalId,
        level: ComplianceLevel | None = None,
    ) -> dict[str, Any]:
        """
        Get compliance frameworks for a vertical.

        Args:
            vertical_id: The vertical ID
            level: Filter by compliance level

        Returns:
            Dict with frameworks list
        """
        params = {"level": level} if level else None
        return self._client.request(
            "GET", f"/api/verticals/{vertical_id}/compliance", params=params
        )

    def suggest(self, task: str) -> dict[str, Any]:
        """
        Suggest the best vertical for a task.

        Args:
            task: Task description

        Returns:
            Dict with suggestions ranked by confidence
        """
        return self._client.request("GET", "/api/verticals/suggest", params={"task": task})

    def create_agent(
        self,
        vertical_id: VerticalId,
        name: str,
        model: str | None = None,
        role: str | None = None,
        traits: list[str] | None = None,
        description: str | None = None,
    ) -> dict[str, Any]:
        """
        Create a vertical-specialized agent.

        Args:
            vertical_id: The vertical ID
            name: Agent name
            model: Model to use
            role: Agent role
            traits: Personality traits
            description: Agent description

        Returns:
            Dict with created agent details
        """
        data: dict[str, Any] = {"name": name}
        if model:
            data["model"] = model
        if role:
            data["role"] = role
        if traits:
            data["traits"] = traits
        if description:
            data["description"] = description
        return self._client.request("POST", f"/api/verticals/{vertical_id}/agent", json=data)

    def create_debate(
        self,
        vertical_id: VerticalId,
        topic: str,
        agent_name: str | None = None,
        rounds: int | None = None,
        consensus: ConsensusType | None = None,
        additional_agents: list[str] | None = None,
        context: str | None = None,
    ) -> dict[str, Any]:
        """
        Create a vertical-specialized debate.

        Args:
            vertical_id: The vertical ID
            topic: Debate topic
            agent_name: Primary agent
            rounds: Number of rounds
            consensus: Consensus type
            additional_agents: Additional agents
            context: Context information

        Returns:
            Dict with created debate details
        """
        data: dict[str, Any] = {"topic": topic}
        if agent_name:
            data["agent_name"] = agent_name
        if rounds:
            data["rounds"] = rounds
        if consensus:
            data["consensus"] = consensus
        if additional_agents:
            data["additional_agents"] = additional_agents
        if context:
            data["context"] = context
        return self._client.request("POST", f"/api/verticals/{vertical_id}/debate", json=data)

    def update_config(
        self,
        vertical_id: VerticalId,
        enabled: bool | None = None,
        model_config: dict[str, Any] | None = None,
        default_traits: list[str] | None = None,
        description: str | None = None,
    ) -> dict[str, Any]:
        """
        Update vertical configuration.

        Args:
            vertical_id: The vertical ID
            enabled: Enable/disable vertical
            model_config: Model configuration
            default_traits: Default agent traits
            description: Updated description

        Returns:
            Dict with updated vertical details
        """
        data: dict[str, Any] = {}
        if enabled is not None:
            data["enabled"] = enabled
        if model_config:
            data["model_config"] = model_config
        if default_traits:
            data["default_traits"] = default_traits
        if description:
            data["description"] = description
        return self._client.request("PUT", f"/api/verticals/{vertical_id}/config", json=data)


class AsyncVerticalsAPI:
    """
    Asynchronous Verticals API.

    Example:
        >>> async with AragoraAsyncClient(base_url="https://api.aragora.ai") as client:
        ...     verticals = await client.verticals.list()
        ...     debate = await client.verticals.create_debate("healthcare", topic="...")
    """

    def __init__(self, client: AragoraAsyncClient):
        self._client = client

    async def list(self, keyword: str | None = None) -> dict[str, Any]:
        """List all available verticals."""
        params = {"keyword": keyword} if keyword else None
        return await self._client.request("GET", "/api/verticals", params=params)

    async def get(self, vertical_id: VerticalId) -> dict[str, Any]:
        """Get vertical details."""
        return await self._client.request("GET", f"/api/verticals/{vertical_id}")

    async def get_tools(self, vertical_id: VerticalId) -> dict[str, Any]:
        """Get tools available for a vertical."""
        return await self._client.request("GET", f"/api/verticals/{vertical_id}/tools")

    async def get_compliance(
        self,
        vertical_id: VerticalId,
        level: ComplianceLevel | None = None,
    ) -> dict[str, Any]:
        """Get compliance frameworks for a vertical."""
        params = {"level": level} if level else None
        return await self._client.request(
            "GET", f"/api/verticals/{vertical_id}/compliance", params=params
        )

    async def suggest(self, task: str) -> dict[str, Any]:
        """Suggest the best vertical for a task."""
        return await self._client.request("GET", "/api/verticals/suggest", params={"task": task})

    async def create_agent(
        self,
        vertical_id: VerticalId,
        name: str,
        model: str | None = None,
        role: str | None = None,
        traits: list[str] | None = None,
        description: str | None = None,
    ) -> dict[str, Any]:
        """Create a vertical-specialized agent."""
        data: dict[str, Any] = {"name": name}
        if model:
            data["model"] = model
        if role:
            data["role"] = role
        if traits:
            data["traits"] = traits
        if description:
            data["description"] = description
        return await self._client.request("POST", f"/api/verticals/{vertical_id}/agent", json=data)

    async def create_debate(
        self,
        vertical_id: VerticalId,
        topic: str,
        agent_name: str | None = None,
        rounds: int | None = None,
        consensus: ConsensusType | None = None,
        additional_agents: list[str] | None = None,
        context: str | None = None,
    ) -> dict[str, Any]:
        """Create a vertical-specialized debate."""
        data: dict[str, Any] = {"topic": topic}
        if agent_name:
            data["agent_name"] = agent_name
        if rounds:
            data["rounds"] = rounds
        if consensus:
            data["consensus"] = consensus
        if additional_agents:
            data["additional_agents"] = additional_agents
        if context:
            data["context"] = context
        return await self._client.request("POST", f"/api/verticals/{vertical_id}/debate", json=data)

    async def update_config(
        self,
        vertical_id: VerticalId,
        enabled: bool | None = None,
        model_config: dict[str, Any] | None = None,
        default_traits: list[str] | None = None,
        description: str | None = None,
    ) -> dict[str, Any]:
        """Update vertical configuration."""
        data: dict[str, Any] = {}
        if enabled is not None:
            data["enabled"] = enabled
        if model_config:
            data["model_config"] = model_config
        if default_traits:
            data["default_traits"] = default_traits
        if description:
            data["description"] = description
        return await self._client.request("PUT", f"/api/verticals/{vertical_id}/config", json=data)

File: sdk/typescript/src/namespaces/verticals.ts
/**
 * Verticals Namespace API
 *
 * Provides domain-specific vertical management:
 * - List and configure industry verticals
 * - Get vertical-specific tools and compliance
 * - Create specialized agents and debates
 */

/**
 * Vertical identifier.
 */
export type VerticalId =
  | 'software'
  | 'legal'
  | 'healthcare'
  | 'accounting'
  | 'finance'
  | 'research'
  | 'education'
  | 'marketing'
  | 'operations'
  | 'general';

/**
 * Compliance enforcement level.
 */
export type ComplianceLevel = 'advisory' | 'warning' | 'enforced';

/**
 * Model configuration.
 */
export interface ModelConfig {
  default_model: string;
  fallback_models: string[];
  temperature: number;
  max_tokens: number;
  top_p?: number;
}

/**
 * Tool configuration.
 */
export interface ToolConfig {
  name: string;
  description: string;
  enabled: boolean;
  parameters: Record<string, unknown>;
  requires_auth: boolean;
  rate_limit_per_minute?: number;
}

/**
 * Compliance rule.
 */
export interface ComplianceRule {
  id: string;
  name: string;
  description: string;
  severity: 'info' | 'warning' | 'error' | 'critical';
  check_type: 'regex' | 'keyword' | 'semantic' | 'custom';
  pattern?: string;
  action: 'block' | 'warn' | 'log' | 'redact';
}

/**
 * Compliance framework.
 */
export interface ComplianceFramework {
  framework: string;
  version: string;
  level: ComplianceLevel;
  rules: ComplianceRule[];
  exemptions: string[];
  last_updated: string;
}

/**
 * Vertical summary (list view).
 */
export interface VerticalSummary {
  vertical_id: VerticalId;
  display_name: string;
  description: string;
  expertise_areas: string[];
  tools: string[];
  agent_count: number;
  enabled: boolean;
}

/**
 * Vertical details (full view).
 */
export interface VerticalDetails extends VerticalSummary {
  compliance_frameworks: string[];
  model_config: ModelConfig;
  default_traits: string[];
  created_at: string;
  updated_at: string;
}

/**
 * Vertical suggestion.
 */
export interface VerticalSuggestion {
  vertical_id: VerticalId;
  display_name: string;
  confidence: number;
  reasoning: string;
  matched_keywords: string[];
}

/**
 * Create agent request.
 */
export interface CreateVerticalAgentRequest {
  name: string;
  model?: string;
  role?: string;
  traits?: string[];
  description?: string;
}

/**
 * Create agent response.
 */
export interface CreateVerticalAgentResponse {
  agent_name: string;
  vertical_id: VerticalId;
  model: string;
  role: string;
  created_at: string;
}

/**
 * Create debate request.
 */
export interface CreateVerticalDebateRequest {
  topic: string;
  agent_name?: string;
  rounds?: number;
  consensus?: 'majority' | 'unanimous' | 'weighted';
  additional_agents?: string[];
  context?: string;
}

/**
 * Create debate response.
 */
export interface CreateVerticalDebateResponse {
  debate_id: string;
  vertical_id: VerticalId;
  topic: string;
  agents: string[];
  rounds: number;
  status: 'created' | 'running' | 'completed';
  created_at: string;
}

/**
 * Update vertical config request.
 */
export interface UpdateVerticalConfigRequest {
  enabled?: boolean;
  model_config?: Partial<ModelConfig>;
  default_traits?: string[];
  description?: string;
}

/**
 * Client interface for verticals operations.
 */
interface VerticalsClientInterface {
  request<T = unknown>(
    method: string,
    path: string,
    options?: { params?: Record<string, unknown>; json?: Record<string, unknown> }
  ): Promise<T>;
}

/**
 * Verticals API namespace.
 *
 * Provides methods for managing industry verticals:
 * - List and configure verticals
 * - Get vertical-specific tools and compliance frameworks
 * - Create specialized agents and debates
 *
 * @example
 * ```typescript
 * const client = createClient({ baseUrl: 'https://api.aragora.ai' });
 *
 * // List all verticals
 * const { verticals } = await client.verticals.list();
 *
 * // Get vertical-specific tools
 * const { tools } = await client.verticals.getTools('legal');
 *
 * // Suggest best vertical for a task
 * const suggestions = await client.verticals.suggest('Review this contract');
 *
 * // Create a vertical-specialized debate
 * const debate = await client.verticals.createDebate('healthcare', {
 *   topic: 'Evaluate treatment options',
 *   rounds: 3,
 * });
 * ```
 */
export class VerticalsAPI {
  constructor(private client: VerticalsClientInterface) {}

  /**
   * List all available verticals.
   */
  async list(options?: { keyword?: string }): Promise<{ verticals: VerticalSummary[] }> {
    return this.client.request('GET', '/api/verticals', {
      params: options as unknown as Record<string, unknown>,
    });
  }

  /**
   * Get vertical details.
   */
  async get(verticalId: VerticalId): Promise<VerticalDetails> {
    return this.client.request('GET', `/api/verticals/${encodeURIComponent(verticalId)}`);
  }

  /**
   * Get tools available for a vertical.
   */
  async getTools(verticalId: VerticalId): Promise<{ tools: ToolConfig[] }> {
    return this.client.request(
      'GET',
      `/api/verticals/${encodeURIComponent(verticalId)}/tools`
    );
  }

  /**
   * Get compliance frameworks for a vertical.
   */
  async getCompliance(
    verticalId: VerticalId,
    options?: { level?: ComplianceLevel }
  ): Promise<{ frameworks: ComplianceFramework[] }> {
    return this.client.request(
      'GET',
      `/api/verticals/${encodeURIComponent(verticalId)}/compliance`,
      {
        params: options as unknown as Record<string, unknown>,
      }
    );
  }

  /**
   * Suggest the best vertical for a task.
   */
  async suggest(task: string): Promise<{ suggestions: VerticalSuggestion[] }> {
    return this.client.request('GET', '/api/verticals/suggest', {
      params: { task },
    });
  }

  /**
   * Create a vertical-specialized agent.
   */
  async createAgent(
    verticalId: VerticalId,
    request: CreateVerticalAgentRequest
  ): Promise<CreateVerticalAgentResponse> {
    return this.client.request(
      'POST',
      `/api/verticals/${encodeURIComponent(verticalId)}/agent`,
      {
        json: request as unknown as Record<string, unknown>,
      }
    );
  }

  /**
   * Create a vertical-specialized debate.
   */
  async createDebate(
    verticalId: VerticalId,
    request: CreateVerticalDebateRequest
  ): Promise<CreateVerticalDebateResponse> {
    return this.client.request(
      'POST',
      `/api/verticals/${encodeURIComponent(verticalId)}/debate`,
      {
        json: request as unknown as Record<string, unknown>,
      }
    );
  }

  /**
   * Update vertical configuration.
   */
  async updateConfig(
    verticalId: VerticalId,
    updates: UpdateVerticalConfigRequest
  ): Promise<VerticalDetails> {
    return this.client.request(
      'PUT',
      `/api/verticals/${encodeURIComponent(verticalId)}/config`,
      {
        json: updates as unknown as Record<string, unknown>,
      }
    );
  }
}
